override entrypoint to env during assemble

bparees opened the pull request:
fixes https://github.com/openshift/source-to-image/issues/475

this basically reimplements https://github.com/openshift/source-to-image/pull/476 but ensures the entrypoint in the final committed image is not changed to "env".

gabemontero commented:
ahh ... more merge pain for me :-)

php-coder commented on pkg/run/run.go:
```diff
@@ -44,14 +44,15 @@ func (b *DockerRunner) Run(config *api.Config) error {
 
 	opts := docker.RunContainerOptions{
 		Image:        config.Tag,
+		Entrypoint:   []string{"/bin/env"},
```
If we always (?) use `/bin/env`, then maybe make it default? Or it's impossible/undesirable?

(BTW I've never heard about `/bin/env` before and in many examples  `/usr/bin/env` is used.)

bparees replied:
make it a default where?

but now i'm wondering if doing this is also going to break images that don't have /bin/env for one reason or another.  (this originally came from @mfojtik, i'm just re-implementing his change because the original one had to be reverted)

php-coder replied:
>make it a default where?

In the methods that are using this (docker.CommitContainer/docker.RunContainer).

php-coder commented on pkg/docker/docker.go:
```diff
@@ -287,6 +291,15 @@ func (d *stiDocker) GetImageWorkdir(name string) (string, error) {
+func (d *stiDocker) GetImageEntrypoint(name string) ([]string, error) {
+	image, err := d.client.InspectImage(name)
+	if err != nil {
+		return nil, err
```
In most of functions we're throwing `errors.NewInspectImageError()` but in some we don't do it. Do you know why? Do we have some rule about it that I don't know?

bparees replied:
I don't know...i'll have to look into how we're using that error to see if we should convert the rest of the cases to also throw that.

php-coder commented on pkg/build/strategies/sti/postexecutorstep.go:
```diff
@@ -114,7 +114,19 @@ func (step *commitImageStep) execute(ctx *postExecutorStepContext) error {
+		return fmt.Errorf("Couldn't get entrypoint of %q image: %v", step.image, err)
+	}
+
+	ctx.imageID, err = commitContainer(step.docker, ctx.containerID, cmd, user, step.builder.config.Tag, step.builder.env, entrypoint, ctx.labels)
```
This method has unit test and would be great if its coverage won't go down :-)

bparees replied:
is that a request to add tests? :)

php-coder replied:
It was a wish to see an about 5 lines in the the existing unit test ;-)

bparees replied:
"see an about"?

On Jul 15, 2016 5:16 PM, "Vyacheslav Semushin" <notifications@github.com>
wrote:

> In pkg/build/strategies/sti/postexecutorstep.go
> <https://github.com/openshift/source-to-image/pull/543#discussion_r71041988>
> :
>
> > @@ -114,7 +114,19 @@ func (step *commitImageStep) execute(ctx *postExecutorStepContext) error {
> >
> >  	ctx.labels = createLabelsForResultingImage(step.builder, step.docker, step.image)
> >
> > -	ctx.imageID, err = commitContainer(step.docker, ctx.containerID, cmd, user, step.builder.config.Tag, step.builder.env, ctx.labels)
> > +	// Set the image entrypoint back to its original value on commit, the running
> > +	// container has "env" as its entrypoint and we don't want to commit that.
> > +	entrypoint, err := step.docker.GetImageEntrypoint(step.image)
> > +	// If the image has no explicit entrypoint, set it to an empty array
> > +	// so we don't default to leaving the entrypoint as "env" upon commit.
> > +	if entrypoint == nil {
> > +		entrypoint = []string{}
> > +	}
> > +	if err != nil {
> > +		return fmt.Errorf("Couldn't get entrypoint of %q image: %v", step.image, err)
> > +	}
> > +
> > +	ctx.imageID, err = commitContainer(step.docker, ctx.containerID, cmd, user, step.builder.config.Tag, step.builder.env, entrypoint, ctx.labels)
>
> It was a wish to see an about 5 lines in the the existing unit test ;-)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/source-to-image/pull/543/files/3ba3ee2654fdbff83ee86ee435cb85bcc3be96f0#r71041988>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEvl3gt4h-lwKTBrLqxwyOrLUnL9dsVKks5qV_iMgaJpZM4JNuCH>
> .
>

php-coder replied:
> "see an about"?

I mean that I'm expecting that there is no more than 5 lines of code to be added.

bparees commented:
@gabemontero yeah, sorry :)  Consider it "motivation to finish the migration to docker-engine sooner"

gabemontero commented:
yep :-) .... i'm actually down to the unit tests and replacing the fake
client stuff ... shouldn't be too much longer

On Fri, Jul 15, 2016 at 4:50 PM, Ben Parees <notifications@github.com>
wrote:

> @gabemontero <https://github.com/gabemontero> yeah, sorry :) Consider it
> "motivation to finish the migration to docker-engine sooner"
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/source-to-image/pull/543#issuecomment-233066428>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADbadLuDZ0hQTRX_XvPY65tQi7Tq-9kKks5qV_KSgaJpZM4JNuCH>
> .
>

bparees commented:
@php-coder i've updated the unit test and added a constant for the entrypoint.

i'm still not sure what to do about the fact that this introduces a requirement that all images (including runtime images) have a "/bin/env" binary available. For the most part i'm not worried about it, but if someone wants to do an extended build with a minimal target image, they can't do it here.

maybe i'll update the code to only set(override) the entrypoint if your image already sets an entrypoint.

php-coder commented:
>if someone wants to do an extended build with a minimal target image, they can't do it here

Is it possible to handle such a case like we're handling case of missing sh/tar? As I understand because we're always run assemble script as `env assemble` it will fail with `env not found`, right? And we can catch it in the same way.

(I don't have an answer, just suggesting idea and I'm not sure that it's possible..)

php-coder commented:
@bparees

>i've updated the unit test and added a constant for the entrypoint.

... but looks like you forgot to push your changes, did you? :)

bparees commented:
actually I did push, but what i didn't do was commit them :)

but not worth you looking now, i'm going to update again to do what i said (not set the entrypoint unless the image explicitly declares one)

bparees commented:
@php-coder ok, ptal again.

php-coder commented on pkg/docker/docker.go:
```diff
@@ -138,13 +139,17 @@ type PullResult struct {
+	// Entrypoint will be used to override the default entrypoint
+	// for the image if it has one.  If the image has no entrypoint,
+	// this value is ignored.
+	Entrypoint       []string
```
What if later we'll want to always set the `Entrypoint`? What we'll do then?

bparees replied:
we'll change the behavior and/or introduce another field w/ that behavior?

mfojtik replied:
+1 (i think the godoc is enough)

mfojtik commented:
@bparees are we setting the entrypoint back to the original value after the assemble finish to run?

bparees commented:
> @bparees are we setting the entrypoint back to the original value after the assemble finish to run?

@mfojtik yes, that's the primary new functionality over your original PR for this issue.

mfojtik commented:
@bparees at that time I thought `/bin/env` must be enough for everybody :-)

mfojtik commented on pkg/build/strategies/sti/postexecutorstep.go:
```diff
@@ -114,7 +115,19 @@ func (step *commitImageStep) execute(ctx *postExecutorStepContext) error {
+	if entrypoint == nil {
+		entrypoint = []string{}
+	}
+	if err != nil {
```
move this up, no point of setting the entrypoint if you get the error.

bparees replied:
good catch

mfojtik commented on pkg/run/run.go:
```diff
@@ -11,6 +11,9 @@ import (
 )
 
+// DefaultEntrypoint is the default entry point used when starting containers
+const DefaultEntrypoint = "/bin/env"
```
`var DefaultEntrypoint = []string{"/bin/env"}` ?

(you don't have to create the arrays when calling this)

bparees replied:
good idea.

mfojtik commented:
minor suggestions, but LGTM!

bparees commented:
@mfojtik updated, will let @php-coder have a final word before merging.

php-coder commented:
LGTM too.

bparees commented:
[merge]
